list.go

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// This file is used to show the list of container checkpoints
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/olekukonko/tablewriter"
+	"github.com/spf13/cobra"
+)
+
+func list(cmd *cobra.Command, args []string) error {
+	allPaths := append([]string{defaultCheckpointPath}, additionalCheckpointPaths...)
+	showTable := false
+
+	var wg sync.WaitGroup
+
+	table := tablewriter.NewWriter(os.Stdout)
+	header := []string{
+		"Namespace",
+		"Pod",
+		"Container",
+		"Time Created",
+		"Checkpoint Name",
+	}
+
+	table.SetHeader(header)
+	table.SetAutoMergeCells(false)
+	table.SetRowLine(true)
+
+	for _, checkpointPath := range allPaths {
+
+		files, err := filepath.Glob(filepath.Join(checkpointPath, "checkpoint-*"))
+		if err != nil {
+			return err
+		}
+
+		if len(files) == 0 {
+			continue
+		}
+
+		showTable = true
+		fmt.Printf("Listing checkpoints in path: %s\n", checkpointPath)
+
+		for _, file := range files {
+			namespace, pod, container, timestamp, err := parseCheckpointName(filepath.Base(file))
```

Comment:
I do not think parsing the file name is a good idea. It was not created to be parse-able and it might change in the future. It probably is faster to parse the name, but we have all the necessary information in the tar archive. Extracting a single file from the archive should give you much easier access to the necessary data.

Reply:
I agree, using data in the tar archive would be more reliable than parsing the file name. We can extract the fields `name` and `checkpointedTime` from `config.dump` (for checkpoints created with CRI-O). The field "name" contains the Kubernetes namespace, pod, and container names.

Reply:
@rst0git @adrianreber as far as I can see that from the config.dump file we can get the container name, image name, image id, ports, network mode, runtime, mounts etc. Although the checkpoint tar file that I am navigating is a podman container's, maybe I need to create a checkpoint of kubernetes pod.

Reply:
@Parthiba-Hazra Podman and CRI-O store different data in `config.dump`. Here is an example `config.dump` to illustrate this difference:
<details>
<summary>Podman</summary>

```json
{
  "spec": {
    "ociVersion": "1.1.0+dev",
    "process": {
      "user": {
        "uid": 0,
        "gid": 0
      },
      "args": [
        "/bin/sh",
        "-c",
        "i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done"
      ],
      "env": [
        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "container=podman"
      ],
      "cwd": "/",
      "capabilities": {
        "bounding": [
          "CAP_CHOWN",
          "CAP_DAC_OVERRIDE",
          "CAP_FOWNER",
          "CAP_FSETID",
          "CAP_KILL",
          "CAP_NET_BIND_SERVICE",
          "CAP_SETFCAP",
          "CAP_SETGID",
          "CAP_SETPCAP",
          "CAP_SETUID",
          "CAP_SYS_CHROOT"
        ],
        "effective": [
          "CAP_CHOWN",
          "CAP_DAC_OVERRIDE",
          "CAP_FOWNER",
          "CAP_FSETID",
          "CAP_KILL",
          "CAP_NET_BIND_SERVICE",
          "CAP_SETFCAP",
          "CAP_SETGID",
          "CAP_SETPCAP",
          "CAP_SETUID",
          "CAP_SYS_CHROOT"
        ],
        "permitted": [
          "CAP_CHOWN",
          "CAP_DAC_OVERRIDE",
          "CAP_FOWNER",
          "CAP_FSETID",
          "CAP_KILL",
          "CAP_NET_BIND_SERVICE",
          "CAP_SETFCAP",
          "CAP_SETGID",
          "CAP_SETPCAP",
          "CAP_SETUID",
          "CAP_SYS_CHROOT"
        ]
      },
      "rlimits": [
        {
          "type": "RLIMIT_NPROC",
          "hard": 4194304,
          "soft": 4194304
        }
      ]
    },
    "root": {
      "path": "rootfs"
    },
    "mounts": [
      {
        "destination": "/proc",
        "type": "proc",
        "source": "proc",
        "options": [
          "nosuid",
          "noexec",
          "nodev"
        ]
      },
      {
        "destination": "/dev",
        "type": "tmpfs",
        "source": "tmpfs",
        "options": [
          "nosuid",
          "strictatime",
          "mode=755",
          "size=65536k"
        ]
      },
      {
        "destination": "/dev/pts",
        "type": "devpts",
        "source": "devpts",
        "options": [
          "nosuid",
          "noexec",
          "newinstance",
          "ptmxmode=0666",
          "mode=0620",
          "gid=5"
        ]
      },
      {
        "destination": "/dev/mqueue",
        "type": "mqueue",
        "source": "mqueue",
        "options": [
          "nosuid",
          "noexec",
          "nodev"
        ]
      },
      {
        "destination": "/sys",
        "type": "sysfs",
        "source": "sysfs",
        "options": [
          "nosuid",
          "noexec",
          "nodev",
          "ro"
        ]
      },
      {
        "destination": "/sys/fs/cgroup",
        "type": "cgroup",
        "source": "cgroup",
        "options": [
          "rprivate",
          "nosuid",
          "noexec",
          "nodev",
          "relatime",
          "ro"
        ]
      }
    ],
    "linux": {
      "sysctl": {
        "net.ipv4.ping_group_range": "0 0"
      },
      "resources": {
        "devices": [
          {
            "allow": false,
            "access": "rwm"
          }
        ],
        "pids": {
          "limit": 2048
        }
      },
      "namespaces": [
        {
          "type": "pid"
        },
        {
          "type": "network"
        },
        {
          "type": "ipc"
        },
        {
          "type": "uts"
        },
        {
          "type": "mount"
        },
        {
          "type": "cgroup"
        }
      ],
      "seccomp": {
        "defaultAction": "SCMP_ACT_ERRNO",
        "defaultErrnoRet": 38,
        "architectures": [
          "SCMP_ARCH_X86_64",
          "SCMP_ARCH_X86",
          "SCMP_ARCH_X32"
        ],
        "syscalls": [
          {
            "names": [
              "bdflush",
              "io_pgetevents",
              "kexec_file_load",
              "kexec_load",
              "migrate_pages",
              "move_pages",
              "nfsservctl",
              "nice",
              "oldfstat",
              "oldlstat",
              "oldolduname",
              "oldstat",
              "olduname",
              "pciconfig_iobase",
              "pciconfig_read",
              "pciconfig_write",
              "sgetmask",
              "ssetmask",
              "swapcontext",
              "swapoff",
              "swapon",
              "sysfs",
              "uselib",
              "userfaultfd",
              "ustat",
              "vm86",
              "vm86old",
              "vmsplice"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "_llseek",
              "_newselect",
              "accept",
              "accept4",
              "access",
              "adjtimex",
              "alarm",
              "bind",
              "brk",
              "capget",
              "capset",
              "chdir",
              "chmod",
              "chown",
              "chown32",
              "clock_adjtime",
              "clock_adjtime64",
              "clock_getres",
              "clock_getres_time64",
              "clock_gettime",
              "clock_gettime64",
              "clock_nanosleep",
              "clock_nanosleep_time64",
              "clone",
              "clone3",
              "close",
              "close_range",
              "connect",
              "copy_file_range",
              "creat",
              "dup",
              "dup2",
              "dup3",
              "epoll_create",
              "epoll_create1",
              "epoll_ctl",
              "epoll_ctl_old",
              "epoll_pwait",
              "epoll_pwait2",
              "epoll_wait",
              "epoll_wait_old",
              "eventfd",
              "eventfd2",
              "execve",
              "execveat",
              "exit",
              "exit_group",
              "faccessat",
              "faccessat2",
              "fadvise64",
              "fadvise64_64",
              "fallocate",
              "fanotify_mark",
              "fchdir",
              "fchmod",
              "fchmodat",
              "fchown",
              "fchown32",
              "fchownat",
              "fcntl",
              "fcntl64",
              "fdatasync",
              "fgetxattr",
              "flistxattr",
              "flock",
              "fork",
              "fremovexattr",
              "fsconfig",
              "fsetxattr",
              "fsmount",
              "fsopen",
              "fspick",
              "fstat",
              "fstat64",
              "fstatat64",
              "fstatfs",
              "fstatfs64",
              "fsync",
              "ftruncate",
              "ftruncate64",
              "futex",
              "futex_time64",
              "futimesat",
              "get_mempolicy",
              "get_robust_list",
              "get_thread_area",
              "getcpu",
              "getcwd",
              "getdents",
              "getdents64",
              "getegid",
              "getegid32",
              "geteuid",
              "geteuid32",
              "getgid",
              "getgid32",
              "getgroups",
              "getgroups32",
              "getitimer",
              "getpeername",
              "getpgid",
              "getpgrp",
              "getpid",
              "getppid",
              "getpriority",
              "getrandom",
              "getresgid",
              "getresgid32",
              "getresuid",
              "getresuid32",
              "getrlimit",
              "getrusage",
              "getsid",
              "getsockname",
              "getsockopt",
              "gettid",
              "gettimeofday",
              "getuid",
              "getuid32",
              "getxattr",
              "inotify_add_watch",
              "inotify_init",
              "inotify_init1",
              "inotify_rm_watch",
              "io_cancel",
              "io_destroy",
              "io_getevents",
              "io_setup",
              "io_submit",
              "ioctl",
              "ioprio_get",
              "ioprio_set",
              "ipc",
              "keyctl",
              "kill",
              "landlock_add_rule",
              "landlock_create_ruleset",
              "landlock_restrict_self",
              "lchown",
              "lchown32",
              "lgetxattr",
              "link",
              "linkat",
              "listen",
              "listxattr",
              "llistxattr",
              "lremovexattr",
              "lseek",
              "lsetxattr",
              "lstat",
              "lstat64",
              "madvise",
              "mbind",
              "membarrier",
              "memfd_create",
              "memfd_secret",
              "mincore",
              "mkdir",
              "mkdirat",
              "mknod",
              "mknodat",
              "mlock",
              "mlock2",
              "mlockall",
              "mmap",
              "mmap2",
              "mount",
              "mount_setattr",
              "move_mount",
              "mprotect",
              "mq_getsetattr",
              "mq_notify",
              "mq_open",
              "mq_timedreceive",
              "mq_timedreceive_time64",
              "mq_timedsend",
              "mq_timedsend_time64",
              "mq_unlink",
              "mremap",
              "msgctl",
              "msgget",
              "msgrcv",
              "msgsnd",
              "msync",
              "munlock",
              "munlockall",
              "munmap",
              "name_to_handle_at",
              "nanosleep",
              "newfstatat",
              "open",
              "open_tree",
              "openat",
              "openat2",
              "pause",
              "pidfd_getfd",
              "pidfd_open",
              "pidfd_send_signal",
              "pipe",
              "pipe2",
              "pivot_root",
              "pkey_alloc",
              "pkey_free",
              "pkey_mprotect",
              "poll",
              "ppoll",
              "ppoll_time64",
              "prctl",
              "pread64",
              "preadv",
              "preadv2",
              "prlimit64",
              "process_mrelease",
              "process_vm_readv",
              "process_vm_writev",
              "pselect6",
              "pselect6_time64",
              "ptrace",
              "pwrite64",
              "pwritev",
              "pwritev2",
              "read",
              "readahead",
              "readdir",
              "readlink",
              "readlinkat",
              "readv",
              "reboot",
              "recv",
              "recvfrom",
              "recvmmsg",
              "recvmmsg_time64",
              "recvmsg",
              "remap_file_pages",
              "removexattr",
              "rename",
              "renameat",
              "renameat2",
              "restart_syscall",
              "rmdir",
              "rseq",
              "rt_sigaction",
              "rt_sigpending",
              "rt_sigprocmask",
              "rt_sigqueueinfo",
              "rt_sigreturn",
              "rt_sigsuspend",
              "rt_sigtimedwait",
              "rt_sigtimedwait_time64",
              "rt_tgsigqueueinfo",
              "sched_get_priority_max",
              "sched_get_priority_min",
              "sched_getaffinity",
              "sched_getattr",
              "sched_getparam",
              "sched_getscheduler",
              "sched_rr_get_interval",
              "sched_rr_get_interval_time64",
              "sched_setaffinity",
              "sched_setattr",
              "sched_setparam",
              "sched_setscheduler",
              "sched_yield",
              "seccomp",
              "select",
              "semctl",
              "semget",
              "semop",
              "semtimedop",
              "semtimedop_time64",
              "send",
              "sendfile",
              "sendfile64",
              "sendmmsg",
              "sendmsg",
              "sendto",
              "set_mempolicy",
              "set_robust_list",
              "set_thread_area",
              "set_tid_address",
              "setfsgid",
              "setfsgid32",
              "setfsuid",
              "setfsuid32",
              "setgid",
              "setgid32",
              "setgroups",
              "setgroups32",
              "setitimer",
              "setns",
              "setpgid",
              "setpriority",
              "setregid",
              "setregid32",
              "setresgid",
              "setresgid32",
              "setresuid",
              "setresuid32",
              "setreuid",
              "setreuid32",
              "setrlimit",
              "setsid",
              "setsockopt",
              "setuid",
              "setuid32",
              "setxattr",
              "shmat",
              "shmctl",
              "shmdt",
              "shmget",
              "shutdown",
              "sigaction",
              "sigaltstack",
              "signal",
              "signalfd",
              "signalfd4",
              "sigpending",
              "sigprocmask",
              "sigreturn",
              "sigsuspend",
              "socket",
              "socketcall",
              "socketpair",
              "splice",
              "stat",
              "stat64",
              "statfs",
              "statfs64",
              "statx",
              "symlink",
              "symlinkat",
              "sync",
              "sync_file_range",
              "syncfs",
              "syscall",
              "sysinfo",
              "syslog",
              "tee",
              "tgkill",
              "time",
              "timer_create",
              "timer_delete",
              "timer_getoverrun",
              "timer_gettime",
              "timer_gettime64",
              "timer_settime",
              "timer_settime64",
              "timerfd",
              "timerfd_create",
              "timerfd_gettime",
              "timerfd_gettime64",
              "timerfd_settime",
              "timerfd_settime64",
              "times",
              "tkill",
              "truncate",
              "truncate64",
              "ugetrlimit",
              "umask",
              "umount",
              "umount2",
              "uname",
              "unlink",
              "unlinkat",
              "unshare",
              "utime",
              "utimensat",
              "utimensat_time64",
              "utimes",
              "vfork",
              "wait4",
              "waitid",
              "waitpid",
              "write",
              "writev"
            ],
            "action": "SCMP_ACT_ALLOW"
          },
          {
            "names": [
              "personality"
            ],
            "action": "SCMP_ACT_ALLOW",
            "args": [
              {
                "index": 0,
                "value": 0,
                "op": "SCMP_CMP_EQ"
              }
            ]
          },
          {
            "names": [
              "personality"
            ],
            "action": "SCMP_ACT_ALLOW",
            "args": [
              {
                "index": 0,
                "value": 8,
                "op": "SCMP_CMP_EQ"
              }
            ]
          },
          {
            "names": [
              "personality"
            ],
            "action": "SCMP_ACT_ALLOW",
            "args": [
              {
                "index": 0,
                "value": 131072,
                "op": "SCMP_CMP_EQ"
              }
            ]
          },
          {
            "names": [
              "personality"
            ],
            "action": "SCMP_ACT_ALLOW",
            "args": [
              {
                "index": 0,
                "value": 131080,
                "op": "SCMP_CMP_EQ"
              }
            ]
          },
          {
            "names": [
              "personality"
            ],
            "action": "SCMP_ACT_ALLOW",
            "args": [
              {
                "index": 0,
                "value": 4294967295,
                "op": "SCMP_CMP_EQ"
              }
            ]
          },
          {
            "names": [
              "arch_prctl"
            ],
            "action": "SCMP_ACT_ALLOW"
          },
          {
            "names": [
              "modify_ldt"
            ],
            "action": "SCMP_ACT_ALLOW"
          },
          {
            "names": [
              "open_by_handle_at"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "bpf",
              "fanotify_init",
              "lookup_dcookie",
              "perf_event_open",
              "quotactl",
              "setdomainname",
              "sethostname",
              "setns"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "chroot"
            ],
            "action": "SCMP_ACT_ALLOW"
          },
          {
            "names": [
              "delete_module",
              "finit_module",
              "init_module",
              "query_module"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "acct"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "kcmp",
              "process_madvise"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "ioperm",
              "iopl"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "clock_settime",
              "clock_settime64",
              "settimeofday",
              "stime"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "vhangup"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 1
          },
          {
            "names": [
              "socket"
            ],
            "action": "SCMP_ACT_ERRNO",
            "errnoRet": 22,
            "args": [
              {
                "index": 0,
                "value": 16,
                "op": "SCMP_CMP_EQ"
              },
              {
                "index": 2,
                "value": 9,
                "op": "SCMP_CMP_EQ"
              }
            ]
          },
          {
            "names": [
              "socket"
            ],
            "action": "SCMP_ACT_ALLOW",
            "args": [
              {
                "index": 2,
                "value": 9,
                "op": "SCMP_CMP_NE"
              }
            ]
          },
          {
            "names": [
              "socket"
            ],
            "action": "SCMP_ACT_ALLOW",
            "args": [
              {
                "index": 0,
                "value": 16,
                "op": "SCMP_CMP_NE"
              }
            ]
          },
          {
            "names": [
              "socket"
            ],
            "action": "SCMP_ACT_ALLOW",
            "args": [
              {
                "index": 2,
                "value": 9,
                "op": "SCMP_CMP_NE"
              }
            ]
          }
        ]
      },
      "maskedPaths": [
        "/proc/acpi",
        "/proc/kcore",
        "/proc/keys",
        "/proc/latency_stats",
        "/proc/timer_list",
        "/proc/timer_stats",
        "/proc/sched_debug",
        "/proc/scsi",
        "/sys/firmware",
        "/sys/fs/selinux",
        "/sys/dev/block"
      ],
      "readonlyPaths": [
        "/proc/asound",
        "/proc/bus",
        "/proc/fs",
        "/proc/irq",
        "/proc/sys",
        "/proc/sysrq-trigger"
      ]
    }
  },
  "id": "1a92a6937e4be4fdd64832792bbd5fb4a92f96a3f95be19a8d48a6272068ec06",
  "name": "looper",
  "lockID": 16,
  "CreateCommand": [
    "podman",
    "run",
    "-d",
    "--name",
    "looper",
    "busybox",
    "/bin/sh",
    "-c",
    "i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done"
  ],
  "RawImageName": "busybox",
  "idMappingsOptions": {
    "HostUIDMapping": true,
    "HostGIDMapping": true,
    "UIDMap": null,
    "GIDMap": null,
    "AutoUserNs": false,
    "AutoUserNsOpts": {
      "Size": 0,
      "InitialSize": 0,
      "PasswdFile": "",
      "GroupFile": "",
      "AdditionalUIDMappings": null,
      "AdditionalGIDMappings": null
    }
  },
  "Dependencies": null,
  "rootfsImageID": "9211bbaa0dbd68fed073065eb9f0a6ed00a75090a9235eca2554c62d1e75c58f",
  "rootfsImageName": "docker.io/library/busybox:latest",
  "ShmDir": "/var/lib/containers/storage/overlay-containers/1a92a6937e4be4fdd64832792bbd5fb4a92f96a3f95be19a8d48a6272068ec06/userdata/shm",
  "shmSize": 65536000,
  "shmSizeSystemd": 0,
  "staticDir": "/var/lib/containers/storage/overlay-containers/1a92a6937e4be4fdd64832792bbd5fb4a92f96a3f95be19a8d48a6272068ec06/userdata",
  "mounts": [
    "/var/lib/containers/storage/overlay-containers/1a92a6937e4be4fdd64832792bbd5fb4a92f96a3f95be19a8d48a6272068ec06/userdata/shm"
  ],
  "createWorkingDir": true,
  "secretsPath": "/var/lib/containers/storage/overlay-containers/1a92a6937e4be4fdd64832792bbd5fb4a92f96a3f95be19a8d48a6272068ec06/userdata/secrets",
  "storageOpts": null,
  "passwd": true,
  "privileged": false,
  "ProcessLabel": "system_u:system_r:container_t:s0:c569,c819",
  "MountLabel": "system_u:object_r:container_file_t:s0:c569,c819",
  "label_nested": false,
  "createNetNS": true,
  "UseImageResolvConf": false,
  "UseImageHosts": false,
  "newNetworks": {
    "podman": {
      "interface_name": "eth0"
    }
  },
  "networkMode": "bridge",
  "command": [
    "/bin/sh",
    "-c",
    "i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done"
  ],
  "stopSignal": 15,
  "stopTimeout": 10,
  "createdTime": "2024-01-19T09:35:21.429939638Z",
  "cgroupManager": "systemd",
  "cgroupsMode": "enabled",
  "cgroupParent": "machine.slice",
  "kubeExitCodePropagation": 0,
  "logPath": "",
  "logTag": "",
  "logSize": 0,
  "logDriver": "journald",
  "conmonPidFile": "/run/containers/storage/overlay-containers/1a92a6937e4be4fdd64832792bbd5fb4a92f96a3f95be19a8d48a6272068ec06/userdata/conmon.pid",
  "postConfigureNetNS": false,
  "runtime": "crun",
  "pause": false,
  "isService": false,
  "sdnotifyMode": "container",
  "healthcheck": null,
  "healthcheck_on_failure_action": 0,
  "umask": "0022",
  "pid_file": "/run/containers/storage/overlay-containers/1a92a6937e4be4fdd64832792bbd5fb4a92f96a3f95be19a8d48a6272068ec06/userdata/pidfile",
  "mountAllDevices": false,
  "readWriteTmpfs": false
}
```
</details>

<details>
<summary>CRI-O</summary>

```json
{
  "id": "f3d1eb1107d2b78de5e445303b753569a814d9973443cc8e792580997f11c352",
  "name": "k8s_counter_counter-gvb5j_default_ee0dc0b9-d4de-4641-9f3f-554b72d4390e_0",
  "rootfsImage": "8c811b4aec35f259572d0f79207bc0678df4c736eeec50bc9fec37ed936a472a",
  "rootfsImageRef": "8c811b4aec35f259572d0f79207bc0678df4c736eeec50bc9fec37ed936a472a",
  "rootfsImageName": "docker.io/library/busybox:1.28",
  "runtime": "runc",
  "createdTime": "2023-10-11T09:36:24.968326678Z",
  "checkpointedTime": "2023-10-12T11:01:57.051768645+01:00",
  "restoredTime": "0001-01-01T00:00:00Z",
  "restored": false
}
```
</details>

Podman (similar to Docker) is mainly used to create and manage containers and it might be more useful for the `checkpointctl list` command to show an overview of container checkpoints that have not been exported, i.e., stored in `/var/lib/containers/storage/overlay-containers/`. Currently, there is no easy way of finding out what checkpoints have been created other than 

```
ls /var/lib/containers/storage/overlay-containers/*/userdata/checkpoint
```


On the other hand, CRI-O is similar to [containerd](https://github.com/containerd/containerd). With CRI-O, the `checkpointctl list` command is somewhat similar to `checkpointctl show /var/lib/kubelet/checkpoints/*`. The difference is that it shows information about the Kubernetes namespace and Pod for each checkpoint, and we can extend the list command with options to filter the results. This is particularly useful when container checkpoints are created automatically.

Reply:
@rst0git thanks for the clarification, I also thought the same. Actually I was having some trouble to install CRI-O, that's why I was not tested that.

Reply:
@adrianreber @rst0git I need one help from you. 
I started minikube using crio runtime -
```
➜ minikube start --container-runtime=crio --feature-gates=ContainerCheckpoint=true
```
applied a demo pod - 
```
➜ kubectl apply -f demo-pod.yaml
```
demo-pod.yaml- 
```
apiVersion: v1
kind: Pod
metadata:
  name: busybox-pod
spec:
  containers:
  - name: busybox-container
    image: busybox
    command: ["sleep", "3600"]
```
but creating a checkpoint was not successful - 
```
docker@minikube:~$ sudo crictl checkpoint busybox-container
E0120 15:37:07.591731   12262 remote_runtime.go:1214] "CheckpointContainer from runtime service failed" err="rpc error: code = Unimplemented desc = unknown method CheckpointContainer for service runtime.v1.RuntimeService" containerID="busybox-container"
FATA[0000] checkpointing the container "busybox-container" failed: rpc error: code = Unimplemented desc = unknown method CheckpointContainer for service runtime.v1.RuntimeService 

```

Reply:
also tried - 
```
➜  list-test curl --insecure --cert ~/.minikube/profiles/criupoc/client.crt --key ~/.minikube/profiles/criupoc/client.key -X POST "https://<minikube ip>:10250/checkpoint/default/busybox-pod/busybox-container"
curl: (28) Failed to connect to <minikube ip> port 10250 after 134275 ms: Couldn't connect to server

```

Reply:
this also fail- 
```
docker@minikube:~$ vi client.crt
docker@minikube:~$ vi client.key
docker@minikube:~$ curl -skv --cert ./client.crt --key ./client.key -X POST "https://localhost:10250/checkpoint/default/busybox-pod/busybox-container"
*   Trying 127.0.0.1:10250...
* Connected to localhost (127.0.0.1) port 10250 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS header, Finished (20):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.2 (OUT), TLS header, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, CERT verify (15):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=minikube@1705764396
*  start date: Jan 20 14:26:36 2024 GMT
*  expire date: Jan 19 14:26:36 2025 GMT
*  issuer: CN=minikube-ca@1705764396
*  SSL certificate verify result: self-signed certificate in certificate chain (19), continuing anyway.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* Using Stream ID: 1 (easy handle 0x55ad97ca0e90)
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
> POST /checkpoint/default/busybox-pod/busybox-container HTTP/2
> Host: localhost:10250
> user-agent: curl/7.81.0
> accept: */*
> 
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
< HTTP/2 500 
< content-type: text/plain; charset=utf-8
< content-length: 175
< date: Sat, 20 Jan 2024 17:02:15 GMT
< 
* Connection #0 to host localhost left intact
checkpointing of default/busybox-pod/busybox-container failed (rpc error: code = Unimplemented desc = unknown method CheckpointContainer for service runtime.v1.Runt
```

Reply:
@rst0git need some more time to update this PR, still having some issue to run kubeadm in my ubuntu machine ;(